### PR TITLE
feat: drill by dimension

### PR DESCRIPTION
**Pull request by ZeRego**
Closes: #3891 

### Description:

This feature is available on both explore and dashboards:
- results table - only on metric cell
- charts - only on series with metric
- big number - only on metric values
- table viz - only on metric cell

When drilling down the query should contain:
- dimension selected in the menu
- metric clicked/selected
- explore filters + dashboard filters + one filter per pivot value + dimension value from selected chart series

<a href="https://www.loom.com/share/13ba56a4abc9475b9314684b86321d49">
    <p>Lightdash - drill by dimension ( chart ) - Watch Video</p>
    
  </a>

<a href="https://www.loom.com/share/262ab30f65d9429698f9cb3bc90a6976">
    <p>Lightdash - drill by dimension ( big number ) - Watch Video</p>
    
  </a>

<a href="https://www.loom.com/share/31cc8f25f0e7421b822b4de3272b9b8b">
    <p>Lightdash - drill by dimension ( table viz ) - Watch Video</p>
  </a>

#### 1. Refactor underlying data provider 

- perhaps rename it to something generic that covers all data actions
- replace `filters` prop with `metricQuery` that will cover both underlying data and drill down 
- extract functions to utils and combine with drill down functions if possible

#### 2. Drill down should not be available for series using dimension vs dimension or metric vs metric

This was one of the acceptance criteria for this ticket but decided to leave it out of this PR since it involves refactor/abstraction of the underlying data util function `getDataFromChartClick`.

#### 3. Fix bug where selected cell is on top of the menu options

Fix might be unnecessary depending on solution for the next point.

Screenshot 2022-12-09 at 16 06 54

#### 4. Have a better way to search the dimension to drill by

> There's a way to search for which dimension you want to add (similar to the filter selection component?)

This acceptance criteria was not included in this PR to keep it as simple as possible. There are a few design questions on how this would look and behave. Should we open a modal with an auto-complete input ?

### CODE REVIEW

File: packages/frontend/src/components/UnderlyingData/UnderlyingDataProvider.tsx [hunk 1]
```diff
@@ -101,7 +104,7 @@ export const UnderlyingDataProvider: FC<Props> = ({
 }) => {
     const [config, setConfig] = useState<UnderlyingDataConfig>();
     const [isModalOpen, setModalOpen] = useState<boolean>(false);
-
+    const { data: explore } = useExplore(tableName);
```
Comment (rephus) [inline packages/frontend/src/components/UnderlyingData/UnderlyingDataProvider.tsx:107]: Does it work with on non explorable contexts like dashboards ?

Comment (ZeRego) [reply packages/frontend/src/components/UnderlyingData/UnderlyingDataProvider.tsx:107]: Yes, `useExplore` is just a request to the BE. You might be confused with `useExplorer` which is a deprecated hook to the explorer provider.

Comment (rephus) [reply packages/frontend/src/components/UnderlyingData/UnderlyingDataProvider.tsx:107]: You're totally right

**Comment by rephus** [comment]
image
 
 I think we need to hide hidden fields like Order date

**Comment by rephus** [comment]
I got a funny wide bar when selecting customer id... looks like it is plotting the field as a category ?  [Screencast from 16-12-22 11:35:17.webm](https://user-images.githubusercontent.com/1983672/208081037-b0aa88e3-b257-461f-990b-4b1d79637968.webm)

File: packages/frontend/src/components/UnderlyingData/DrillDownMenuItem.tsx [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        explore &&
+        row &&
+        metricQuery
+    ) {
+        return (
+            <MenuItem2 text="Drill by" icon="path">
+                {Object.values(explore.tables).map((table) => (
+                    <MenuItem2 key={table.name} text={table.label}>
+                        {Object.values(table.dimensions)
+                            .filter((dimension) => !dimension.hidden)
```
Comment (rephus) [inline packages/frontend/src/components/UnderlyingData/DrillDownMenuItem.tsx:174]: It seems this hidden filter is not working..

**Review by rephus — APPROVED** [review]
I tested this and it works pretty well. I just made 2 comments... and the cell covering the popup is very annoying, but I know it is not in scope, so happy to fix this later.

**Comment by ZeRego** [comment]
> I got a funny wide bar when selecting customer id... looks like it is plotting the field as a category ? [Screencast from 16-12-22 11:35:17.webm](https://user-images.githubusercontent.com/1983672/208081037-b0aa88e3-b257-461f-990b-4b1d79637968.webm)

The problem is in `main` as well, opened a ticket for that https://github.com/lightdash/lightdash/issues/3985

File: packages/frontend/src/components/UnderlyingData/DrillDownMenuItem.tsx (hunk 2 above)
Comment (ZeRego) [reply packages/frontend/src/components/UnderlyingData/DrillDownMenuItem.tsx:174]: > I think we need to hide hidden fields like Order date

`Order date` Is not a hidden field. We just have some logic in the sidebar to transform it into a collapsible group.
There is a ticket ( can't find it ) about removing that option from the filters autocomplete as well.

**Pull request merged**